src/storage/src/hummock/shared_buffer/mod.rs

```diff
@@ -17,22 +17,67 @@ pub mod shared_buffer_batch;
 #[allow(dead_code)]
 pub mod shared_buffer_uploader;
 
-use std::collections::HashMap;
+use std::collections::{BTreeMap, HashMap};
 use std::ops::RangeBounds;
 
 use itertools::Itertools;
+use risingwave_hummock_sdk::is_remote_sst_id;
+use risingwave_hummock_sdk::key::user_key;
+use risingwave_pb::hummock::{KeyRange, SstableInfo, VNodeBitmap};
 
 use self::shared_buffer_batch::SharedBufferBatch;
-use crate::hummock::shared_buffer::shared_buffer_batch::IndexedSharedBufferBatches;
 use crate::hummock::shared_buffer::shared_buffer_uploader::{UploadTaskId, UploadTaskPayload};
-use crate::hummock::utils::range_overlap;
+use crate::hummock::utils::{filter_single_sst, range_overlap};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum UncommittedData {
+    Sst(SstableInfo),
+    Batch(SharedBufferBatch),
+}
+
+fn get_sst_key_range(info: &SstableInfo) -> &KeyRange {
+    let key_range = info
+        .key_range
+        .as_ref()
+        .expect("local sstable should have key range");
+    assert!(
+        !key_range.inf,
+        "local sstable should not have infinite key range. Sstable info: {:?}",
+        info,
+    );
+    key_range
+}
+
+impl UncommittedData {
+    pub fn start_user_key(&self) -> &[u8] {
+        match self {
+            UncommittedData::Sst(info) => {
+                let key_range = get_sst_key_range(info);
+                user_key(key_range.left.as_slice())
+            }
+            UncommittedData::Batch(batch) => batch.start_user_key(),
+        }
+    }
+
+    pub fn end_user_key(&self) -> &[u8] {
+        match self {
+            UncommittedData::Sst(info) => {
+                let key_range = get_sst_key_range(info);
+                user_key(key_range.right.as_slice())
+            }
+            UncommittedData::Batch(batch) => batch.end_user_key(),
+        }
+    }
+}
+
+/// `{ (end key) -> batch }`
+pub(crate) type KeyIndexedUncommittedData = BTreeMap<Vec<u8>, UncommittedData>;
 
 #[derive(Default, Debug)]
 pub struct SharedBuffer {
-    non_upload_batches: IndexedSharedBufferBatches,
-    replicate_batches: IndexedSharedBufferBatches,
-    uploading_batches: HashMap<UploadTaskId, IndexedSharedBufferBatches>,
-
+    uncommitted_data: KeyIndexedUncommittedData,
+    replicate_batches: BTreeMap<Vec<u8>, SharedBufferBatch>,
+    uploading_tasks: HashMap<UploadTaskId, KeyIndexedUncommittedData>,
```

Comment:
Shall we store `BTreeMap<Vec<u8>, SharedBufferBatch>` in `uploading_tasks` instead of `KeyIndexedUncommittedData`? It is all shared buffer data and there is no need to wrap in in a enum.

Reply:
Currently in this PR it is all shared buffer data, but for spill to disk, since we can't keep track of the order information in remote meta, we need to compact the local ssts and write batches together. In this scenario, both the ssts and write batches will exist in a upload task